lib/integrations/awsoidc/deployservice_iam_jointoken.go

```diff
@@ -0,0 +1,188 @@
+)
+
+const (
+	// awsSTSShortName is the service name of the AWS STS
+	// Used to build the ARN matcher in the IAM Join Token.
+	awsSTSShortName = "sts"
+)
+
+// TokenService defines the required methods to upsert the Provision Token used by the Deploy Service.
+type TokenService interface {
+	// GetToken returns a provision token by name.
+	GetToken(ctx context.Context, name string) (types.ProvisionToken, error)
+
+	// UpsertToken creates or updates a provision token.
+	UpsertToken(ctx context.Context, token types.ProvisionToken) error
+}
+
+// upsertIAMJoinTokenRequest is the request to create or update a Provision Token that has a rule which allows AWS Identities to join the cluster.
+// Allowed AWS Identities are the ones described
+// It gives access for those identities to join as system roles.
+type upsertIAMJoinTokenRequest struct {
+	// tokenName is the Token's name to create or update.
+	tokenName string
+
+	// accountID is the allowed AWS Account ID.
+	accountID string
+
+	// region is the allowed AWS Region.
+	region string
+
+	// iamRole is the allowed IAM Role.
+	iamRole string
+
+	// deploymentMode is the service configuration that is going to be deployed
+	deploymentMode string
+
+	// awsARN is the IAM Role's ARN.
+	// This value is calculated using the AWS account, region and IAM role.
+	awsARN string
+}
+
+// CheckAndSetDefaults verifies the required fields are present.
+func (u *upsertIAMJoinTokenRequest) CheckAndSetDefaults() error {
+	if u.tokenName == "" {
+		return trace.BadParameter("token name is required")
+	}
+
+	if u.accountID == "" {
+		return trace.BadParameter("accound id is required")
+	}
+
+	if u.region == "" {
+		return trace.BadParameter("region is required")
+	}
+
+	if u.iamRole == "" {
+		return trace.BadParameter("iam role is required")
+	}
+
+	if !slices.Contains(DeploymentModes, u.deploymentMode) {
+		return trace.BadParameter("invalid deployment mode, please use one of the following: %v", DeploymentModes)
+	}
+
+	partition := awsutils.GetPartitionFromRegion(u.region)
+
+	// Example ARN:
+	// arn:aws:sts::278576220453:assumed-role/MarcoTestRoleOIDCProvider/1688143717490080920
+	u.awsARN = arn.ARN{
+		Partition: partition,
+		Service:   awsSTSShortName,
+		AccountID: u.accountID,
+		Resource:  fmt.Sprintf("assumed-role/%s/*", u.iamRole),
+	}.String()
+
+	return nil
+}
+
+// upsertIAMJoinToken creates or updates a Provision Token with a rule that allows Teleport Services with given AWS Identities to join the cluster.
+// The allowed Identities are the ones provided in the request (Account, Region and AWSRole).
+// It gives access for those identities to join as system roles.
+func upsertIAMJoinToken(ctx context.Context, req upsertIAMJoinTokenRequest, clt TokenService) error {
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	rule := &types.TokenRule{
+		AWSAccount: req.accountID,
+		AWSARN:     req.awsARN,
+	}
+
+	token, err := clt.GetToken(ctx, req.tokenName)
+	switch {
```

Comment:
IMO in this case switch makes the flow very weird. My proposal
```go
if trace.IsNotFound(err) {
    token = ....
    err = nil
}

if err != nil {
    return trace.Wrap(err)
}

...

Reply:
It's a bit complex because it has the `fallthrough` down below
I could switch to a plain `if` condition.

Given that you approve, I'll merge as is
But feel free to comment again and I'll try to simplify this part